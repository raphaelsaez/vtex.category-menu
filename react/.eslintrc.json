{
  "extends": "vtex-react",
  "env": {
    "browser": true,
    "node": false
  },
  "rules": {
    "jsx-a11y/click-events-have-key-events": "off",
    "jsx-a11y/no-noninteractive-element-interactions": "off",
    "jsx-a11y/no-static-element-interactions": "off"
  }
}
