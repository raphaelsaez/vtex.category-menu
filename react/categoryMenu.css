.itemContainer {
  max-height: 31.875rem;
}
.itemContainer--department {}
.itemContainer--category {}

.animation {
  left: -0.5rem;
}

@media only screen and (max-width: 40rem) {
  body.sidebarOpen {
      overflow-y: hidden;
  }
}

.container {}

.section--department {}
.section--category {}

.departmentList {}
.departmentLink {}

.firstLevelList {}
.firstLevelLinkContainer {}
.firstLevelLink {}

.secondLevelList {}
.secondLevelLinkContainer {}
.secondLevelLink {}

.submenuList {}
.submenuItem {}

.sidebarOpen {}

.sidebarScrim {}

.sidebar {}

.sidebarHeader {}

.sidebarContent {}

.sidebarItem {}

.sidebarContainer {}

.menuContainer {}

.sidebarItemContainer {}
