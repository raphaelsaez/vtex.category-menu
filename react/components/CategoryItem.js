import React, { useState, useRef } from 'react'
import { Link } from 'vtex.render-runtime'
import PropTypes from 'prop-types'
import { categoryItemShape } from '../propTypes'

import ItemContainer from './ItemContainer'
import classNames from 'classnames'
import styles from '../categoryMenu.css'
import categoryMenuPosition, {
  getMenuPositionValues,
} from '../utils/categoryMenuPosition'

/**
 * Component that represents a single category displayed in the menu, also displays
 * the subcategories, if the provided category has them
 */
const CategoryItem = ({
  department,
  subcategoryLevels,
  menuPosition,
  department: { name, slug },
  noRedirect,
  isCategorySelected,
}) => {
  const [isHover, setHover] = useState(false)
  const itemRef = useRef(null)

  const handleCloseMenu = () => {
    setHover(false)
  }
  menuPosition = 'center'
  const categoryClasses = classNames(
    styles.departmentLink,
    'w-100 pv5 no-underline t-small outline-0 db tc link truncate bb bw1 c-muted-1',
    {
      'b--transparent': !isHover && !isCategorySelected,
      'b--action-primary pointer': isHover || isCategorySelected,
      mr8: menuPosition === categoryMenuPosition.DISPLAY_LEFT.value,
      ml8: menuPosition === categoryMenuPosition.DISPLAY_RIGHT.value,
      mh6: menuPosition === categoryMenuPosition.DISPLAY_CENTER.value,
    }
  )
  menuPosition = 'left'
  const containerStyle = {
    top:
      itemRef.current &&
      itemRef.current.offsetTop + itemRef.current.clientHeight,
    display: isHover ? 'flex' : 'none',
  }

  return (
    <li
      className={`${styles.itemContainer} ${
        styles['itemContainer--department']
      } flex items-center db list`}
      ref={itemRef}
      onMouseEnter={() => setHover(true)}
      onMouseLeave={handleCloseMenu}
    >
      {noRedirect ? (
        <span className={categoryClasses}>{name}</span>
      ) : (
        <Link
          onClick={handleCloseMenu}
          page="store.search#department"
          params={{ department: slug }}
          className={categoryClasses}
        >
          {name}
        </Link>
      )}
      {subcategoryLevels > 0 && department.children.length > 0 && (
        <ItemContainer
          menuPosition={menuPosition}
          containerStyle={containerStyle}
          departments={department.children}
          parentSlug={department.slug}
          onCloseMenu={handleCloseMenu}
          showSecondLevel={subcategoryLevels === 2}
        />
      )}
    </li>
  )
}

CategoryItem.propTypes = {
  /** Department to be displayed */
  department: categoryItemShape.isRequired,
  /** Set use of Link component */
  noRedirect: PropTypes.bool,
  /** Number of subcategory levels */
  subcategoryLevels: PropTypes.oneOf([0, 1, 2]),
  /** Defines the position of the category menu */
  menuPosition: PropTypes.oneOf(getMenuPositionValues()),
  /** Menu category selection */
  isCategorySelected: PropTypes.bool,
}

export default CategoryItem
