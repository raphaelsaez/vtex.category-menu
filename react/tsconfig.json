{
  "extends": "@vtex/tsconfig",
  "include": ["./**/*.ts", "./**/*.tsx", "./typings/*.d.ts"]
}
