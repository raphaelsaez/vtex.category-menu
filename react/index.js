import PropTypes from 'prop-types'
import React, { useState } from 'react'
import { graphql } from 'react-apollo'
import { injectIntl, intlShape } from 'react-intl'
import { IconMenu } from 'vtex.store-icons'
import { compose } from 'ramda'
import classNames from 'classnames'
import { Container } from 'vtex.store-components'

import CategoryItem from './components/CategoryItem'
import SideBar from './components/SideBar'
import { categoryPropType } from './propTypes'
import getCategories from './queries/categoriesQuery.gql'
import Options from './constants/Options'

import styles from './categoryMenu.css'
import categoryMenuPosition, {
  getMenuPositionNames,
  getMenuPositionValues,
} from './utils/categoryMenuPosition'

const DEFAULT_SUBCATEGORIES_LEVELS = 1

/**
 * Component that represents the menu containing the categories of the store
 */
const CategoryMenu = ({
  mobileMode = false,
  showAllDepartments = true,
  showSubcategories = true,
  menuPosition = categoryMenuPosition.DISPLAY_CENTER.value,
  departments = [],
  data: { categories = [] },
  intl,
  manualDepartments = [],
  strategyTypes,
}) => {
  const [sideBarVisible, setSidebarVisible] = useState(false)

  const handleSidebarToggle = () => {
    setSidebarVisible(prevVisible => !prevVisible)
  }

  const departmentsIds = departments.map(dept => dept.id)
  const departmentsSelected = categories.filter(category =>
    departmentsIds.includes(category.id)
  )

  let visibleDepartments =
    (departmentsSelected.length && departmentsSelected) || categories

  switch (strategyTypes) {
    case Options.BOTH:
      visibleDepartments = visibleDepartments.concat(manualDepartments)
      break
    case Options.MANUAL:
      visibleDepartments = manualDepartments
      break
  }

  if (mobileMode) {
    return (
      <div className={`${styles.sidebarContainer} ${styles.mobile}`}>
        <SideBar
          visible={sideBarVisible}
          title={intl.formatMessage({
            id: 'store/category-menu.departments.title',
          })}
          departments={visibleDepartments}
          onClose={handleSidebarToggle}
          showSubcategories={showSubcategories}
        />
        <div className="flex pa4 pointer" onClick={handleSidebarToggle}>
          <IconMenu size={20} />
        </div>
      </div>
    )
  }

  const desktopClasses = classNames(
    `${styles.container} w-100 bg-base dn flex-m`,
    {
      'justify-start': menuPosition === categoryMenuPosition.DISPLAY_LEFT.value,
      'justify-end': menuPosition === categoryMenuPosition.DISPLAY_RIGHT.value,
      'justify-center':
        menuPosition === categoryMenuPosition.DISPLAY_CENTER.value,
    }
  )

  return (
    <nav className={desktopClasses}>
      <Container
        className={`${styles['section--department']} justify-center flex`}
      >
        <ul
          className={`${styles.departmentList} pa0 list ma0 flex flex-wrap flex-row t-action overflow-hidden h3`}
        >
          {showAllDepartments && (
            <CategoryItem
              noRedirect
              menuPosition={menuPosition}
              subcategoryLevels={
                DEFAULT_SUBCATEGORIES_LEVELS + showSubcategories
              }
              department={{
                children: visibleDepartments,
                name: intl.formatMessage({
                  id: 'store/category-menu.departments.title',
                }),
              }}
            />
          )}
        </ul>
      </Container>
    </nav>
  )
}

CategoryMenu.propTypes = {
  /** Categories query data */
  data: PropTypes.shape({
    loading: PropTypes.bool.isRequired,
    categories: PropTypes.arrayOf(categoryPropType),
  }),
  /** Set mobile mode */
  mobileMode: PropTypes.bool,
  /** Whether to show the departments category or not */
  showAllDepartments: PropTypes.bool,
  /** Whether to show subcategories or not */
  showSubcategories: PropTypes.bool,
  /** Defines the position of the category menu */
  menuPosition: PropTypes.oneOf(getMenuPositionValues()),
  /** Intl */
  intl: intlShape,
  /** Departments.js to be shown in the desktop mode. */
  departments: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
    })
  ),
  manualDepartments: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
      slug: PropTypes.string,
      children: PropTypes.arrayOf(
        PropTypes.shape({
          name: PropTypes.string,
          slug: PropTypes.string,
          children: PropTypes.arrayOf(
            PropTypes.shape({
              name: PropTypes.string,
              slug: PropTypes.string,
            })
          ),
        })
      ),
    })
  ),
  strategyTypes: PropTypes.string,
}

CategoryMenu.schema = {
  title: 'admin/editor.category-menu.title',
  description: 'admin/editor.category-menu.description',
  type: 'object',
  properties: {
    showAllDepartments: {
      type: 'boolean',
      title: 'admin/editor.category-menu.show-departments-category.title',
      default: true,
    },
    menuPosition: {
      title: 'admin/editor.category-menu.disposition-type.title',
      type: 'string',
      enum: getMenuPositionValues(),
      enumNames: getMenuPositionNames(),
      default: categoryMenuPosition.DISPLAY_CENTER.value,
      isLayout: true,
    },
    showSubcategories: {
      type: 'boolean',
      title: 'admin/editor.category-menu.show-subcategories.title',
      default: true,
    },
    departments: {
      title: 'store/category-menu.departments.title',
      type: 'array',
      minItems: 0,
      items: {
        title: 'admin/editor.category-menu.departments.items.title',
        type: 'object',
        properties: {
          id: {
            title: 'admin/editor.category-menu.departments.items.id',
            type: 'number',
          },
        },
      },
    },
    strategyTypes: {
      title: 'Strategy?',
      type: 'string',
      enum: ['Both', 'Manual', 'Automatic'],
    },
    manualDepartments: {
      title: 'Departments',
      type: 'array',
      items: {
        properties: {
          name: {
            type: 'string',
            title: "Department's Name",
          },
          slug: {
            type: 'string',
            title: "Department's slug",
          },
          children: {
            title: 'Category',
            type: 'array',
            items: {
              properties: {
                name: {
                  type: 'string',
                  title: "Category's Name",
                },
                slug: {
                  type: 'string',
                  title: "Category's slug",
                },
                children: {
                  type: 'array',
                  title: 'Subcategory',
                  items: {
                    properties: {
                      name: {
                        type: 'string',
                        title: "Subcategory's Name",
                      },
                      slug: {
                        type: 'string',
                        title: "Subcategory's slug",
                      },
                    },
                  },
                },
              },
            },
          },
        },
      },
    },
  },
}

export default compose(
  graphql(getCategories),
  injectIntl
)(CategoryMenu)
