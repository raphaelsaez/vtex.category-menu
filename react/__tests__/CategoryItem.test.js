import React from 'react'
import { render } from '@vtex/test-tools/react'

import CategoryItem from '../components/CategoryItem'

describe('CategoryItem component', () => {
  let wrapper

  beforeEach(() => {
    const categoryMock = {
      id: 1,
      name: 'Category',
      slug: 'category',
      hasChildren: true,
      children: [
        {
          id: 2,
          name: 'Sub-category',
          slug: 'sub-category',
        },
      ],
    }

    wrapper = render(<CategoryItem category={categoryMock} />)
  })

  it('should have the correct href', () => {
    const element = wrapper.getByText(/CATEGORY/i)

    expect(element.href).toMatch(/\/category\/s$/i)
  })

  it('should match snapshot', () => {
    expect(wrapper.asFragment()).toMatchSnapshot()
  })
})
