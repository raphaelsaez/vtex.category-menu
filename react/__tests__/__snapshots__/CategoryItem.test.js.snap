// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryItem component should match snapshot 1`] = `
<DocumentFragment>
  <li
    class="itemContainer itemContainer--department flex items-center db list"
  >
    <a
      class="departmentLink w-100 pv5 no-underline t-small outline-0 db tc link truncate bb bw1 c-muted-1 b--transparent"
      href="/category/s"
    >
      Category
    </a>
  </li>
</DocumentFragment>
`;
